sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -0,0 +1,200 @@
+                nonempty += 1;
+            }
+            if nonempty >= max_addr {
+                frameno |= Self::EOF;
+            } else {
+                let skip_addr = nonempty & !(512 - 1);
+                if skip_addr > addr {
+                    if last_frame_end_addr == 0 {
+                        addr = skip_addr;
+                    } else if last_frame_end_addr & 31 != 0 {
+                        addr = skip_addr;
+                    }
+                }
+            }
+            
+            if addr + Self::DATA_LEN / 4 >= max_addr {
+                frameno |= Self::EOF;
+            }
+
+            let mut frame = Frame {
+                header: FrameHeader {
+                    frame_num: frameno,
+                    flash_offset: (addr * 4 + 65536 * 8) as u32,
+                    ..Default::default()
+                },
+                ..Default::default()
+            };
+            let slice_size = std::cmp::min(Self::DATA_LEN, payload.len() - addr * 4);
+            frame.data[..slice_size].copy_from_slice(&payload[addr * 4..addr * 4 + slice_size]);
+            frame.header.hash = frame.header_hash();
+            frames.push(frame);
+
+            
+            addr += Self::DATA_LEN / 4;
+            frameno += 1;
+            last_frame_end_addr = addr;
+        }
+        frames
+    }
+}
+
+/// Implements the bootstrap protocol of previous Google Titan family chips.
+pub struct Legacy {
+    pub inter_frame_delay: Duration,
+    pub flash_erase_delay: Duration,
+}
+
+impl Legacy {
+    const INTER_FRAME_DELAY: Duration = Duration::from_millis(20);
+    const FLASH_ERASE_DELAY: Duration = Duration::from_millis(200);
+
+    /// Creates a new `Primitive` protocol updater from `options`.
+    pub fn new(options: &BootstrapOptions) -> Self {
+        Self {
+            inter_frame_delay: options.inter_frame_delay.unwrap_or(Self::INTER_FRAME_DELAY),
+            flash_erase_delay: options.flash_erase_delay.unwrap_or(Self::FLASH_ERASE_DELAY),
+        }
+    }
+}
+
+impl UpdateProtocol for Legacy {
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()> {
+        let frames = Frame::from_payload(payload);
+
+        let mut i = 0;
+        while i < frames.len() {
+            let frame = &frames[i];
+            log::info!(
+                "Writing frame {} (offset {:x?})",
+                i,
+                frame.header.flash_offset
+            );
+
+            // Write the frame and read back the hash of the previous frame.
+            let mut prev_hash = [0u8; std::mem::size_of::<Frame>()];
+            spi.run_transaction(&mut [Transfer::Both(frame.as_bytes(), &mut prev_hash)])?;
+
+            std::thread::sleep(self.inter_frame_delay);
+            let want_hash = frames[if i == 0 { 0 } else { i - 1 }].frame_hash();
+            if prev_hash[..Frame::HASH_LEN] != want_hash {
```

Comment:
I believe error codes from the H1D bootstrap implementation are detectable here.  If the hash is the form of: `<byte> <31 zeros>`, then the `byte` can be interpreted as an error code from the chip.

Can you add a TODO to add error reporting and send that in a follow-up PR?

Reply:
I have looked at the code.  32 bytes of identical value seems to be the way that an error is indicated.  I have added code to recognize that case (32 repeated 0x00 or 0xFF results in retries, any other value causes aborting with an Err().)